src/sentry/tasks/unmerge.py

```diff
@@ -0,0 +1,544 @@
+                organization_id=organization_id,
+                version=version,
+            ),
+        ),
+    }
+
+
+def merge_mappings(values):
+    result = {}
+    for value in values:
+        result.update(value)
+    return result
+
+
+initial_fields = {
+    'culprit': lambda event: generate_culprit(
+        event.data,
+        event.platform,
+    ),
+    'data': lambda event: {
+        'last_received': event.data.get('received') or float(event.datetime.strftime('%s')),
+        'type': event.data['type'],
+        'metadata': event.data['metadata'],
+    },
+    'last_seen': lambda event: event.datetime,
+    'level': lambda event: LOG_LEVELS_MAP.get(
+        event.get_tag('level'),
+        logging.ERROR,
+    ),
+    'message': lambda event: event.message,
+    'times_seen': lambda event: 0,
+}
+
+
+backfill_fields = {
+    'platform': lambda caches, data, event: event.platform,
+    'logger': lambda caches, data, event: event.get_tag('logger') or DEFAULT_LOGGER_NAME,
+    'first_seen': lambda caches, data, event: event.datetime,
+    'active_at': lambda caches, data, event: event.datetime,
+    'first_release': lambda caches, data, event: caches[Release](
+        event.project.organization_id,  # TODO
+        event.get_tag('sentry:release'),
+    ) if event.get_tag('sentry:release') else data.get('first_release', None),  # TODO: double check logic.
+    'times_seen': lambda caches, data, event: data['times_seen'] + 1,
+    'score': lambda caches, data, event: ScoreClause.calculate(
+        data['times_seen'] + 1,
+        data['last_seen'],
+    ),
+}
+
+
+def get_group_creation_attributes(caches, events):
+    latest_event = events[0]
+    return reduce(
+        lambda data, event: merge_mappings([
+            data,
+            {name: f(caches, data, event) for name, f in backfill_fields.items()},
+        ]),
+        events,
+        {name: f(latest_event) for name, f in initial_fields.items()},
+    )
+
+
+def get_group_backfill_attributes(caches, group, events):
+    return {
+        k: v for k, v in
+        reduce(
+            lambda data, event: merge_mappings([
+                data,
+                {name: f(caches, data, event) for name, f in backfill_fields.items()},
+            ]),
+            events,
+            {name: getattr(group, name) for name in set(initial_fields.keys()) | set(backfill_fields.keys())},
+        ).items()
+        if k in backfill_fields
+    }
+
+
+def get_fingerprint(event):
+    # TODO: This *might* need to be protected from an IndexError?
+    primary_hash = get_hashes_for_event(event)[0]
+    return md5_from_hash(primary_hash)
+
+
+def migrate_events(caches, project, source_id, destination_id, fingerprints, events, actor_id):
+    # XXX: This is only actually able to create a destination group and migrate
+    # the group hashes if there are events that can be migrated. How do we
+    # handle this if there aren't any events? We can't create a group (there
```

Comment:
> would it ever actually happen that there aren't any events? seems like if we ever do this operation there must be events to migrate, can we not make/rely on that guarantee?

Nothing enforces it, so it's theoretically possible and I'd like to plan for it (or at least note that it's possible, like here.) A short aside: in my opinion it's a good idea to never assume anything is going to work as it "should" unless a more reliable piece of technology enforces that assumed constraint. This is why I advocate so much for maintaining referential integrity through things like foreign key constraints. (I'm also aware that this isn't necessarily a shared opinion of the team.)

The only ways that I can think of this would happen is (note that this list isn't exhaustive, it's just what I could come up with off of the top of my head):

1. All of the events with a specific hash are deleted via cleanup, or some other means. (I'm not even sure if there is an API endpoint for this, off of the top of my head?)
2. [A "secondary" fingerprint could be assigned to a group if it returns multiple hashes](https://github.com/getsentry/sentry/blob/fe03d388e52fe7c73330857c59b0f4f3fd6104ac/src/sentry/event_manager.py#L836-L849) (e.g. application hash and system hash), and since this task only unmerges the first returned ("primary") hash for event (see L136 in this file), if you tried to unmerge the "secondary" hash, nothing would happen.

I don't think it's likely, though, except maybe in that second case. Plus, what would even be migrated, other than just the `GroupHash` record itself? I think we'll probably want to improve the UX here in the future, but I don't have any good ideas off of the top of my head.